Zone11: different handling of root zone, TLD zones and zones in .arpa

marc-vanderwal opened the pull request:
## Purpose

This PR amends the specification for Zone11 so that different message tags of different severity levels are output if Zonemaster is run on the root zone, a TLD zone or a zone inside .arpa. It also cleans up some rough edges and migrates to MethodsV2.

## Context

See #1256.

## Changes

- Root zones, TLD zones and zones inside .arpa are tested differently: an INFO is output if no SPF or a null SPF is found, and a message tag of higher level is output if a non-null SPF is found.
- Use MethodsV2 instead of legacy Methods.
- Miscellaneous small rewordings

### Breaking message tag changes

This PR changes the arguments to the following existing message tags:

| Tag                              | Old arguments | New arguments |
|:---------------------------------|:--------------|:--------------|
| Z11_DIFFERENT_SPF_POLICIES_FOUND | ns_ip_list    | ns_list       |

This PR replaces the following message tags with new tags:

| Old tag                   | Old tag arguments | New tag                  | New tag arguments |
|:--------------------------|:------------------|:-------------------------|:------------------|
| Z11_SPF1_MULTIPLE_RECORDS | ns_ip_list        | Z11_SPF_MULTIPLE_RECORDS | ns_list           |
| Z11_SPF1_SYNTAX_ERROR     | ns_ip_list        | Z11_SPF_SYNTAX_ERROR     | ns_list           |
| Z11_SPF1_SYNTAX_OK        | ns_ip_list        | Z11_SPF_SYNTAX_OK        | ns_list           |

## How to test this PR

Review.

marc-vanderwal committed:
Zone11: handle root, TLDs and .arpa differently

The root zone, TLDs and zones in .arpa are not expected to be domains
appropriate for receiving e-mail. They should not be appropriate as
sender e-mail addresses either.

In the current wording of the specification, a NOTICE (Z11_NO_SPF_FOUND)
is output telling the user that no SPF record was found, when running
Zonemaster on such zones. That message tends to stick out like a sore
thumb, especially on well-configured TLD zones. That NOTICE should not
be there.

The specification is amended so that it outputs specific message tags
when the test is being run on such a zone: if there is no SPF record or
it is a null SPF policy, then we output a special message tag of level
INFO, and if it is a non-null SPF policy, it’s another message of a
higher level.

In doing so, this has the effect of squelching the
NOTICE (Z11_NO_SPF_FOUND) that users have been complaining about when
testing root, TLDs and zones inside .arpa.

marc-vanderwal committed:
Switch Zone11 to MethodsV2

Since we’ve overhauling Zone11, we might as well do some more
maintenance work on the specification.

marc-vanderwal committed:
Zone11: miscellaneous wording changes

This commit implements a few miscellaneous minor wording changes.

Firstly, shorten the msgid of Z11_DIFFERENT_SPF_POLICIES_FOUND in order
to reduce redundant information. This tag is always emitted along with
Z11_INCONSISTENT_SPF_POLICIES.

Secondly, remove explicit mentions to “version 1” of SPF except for the
very first occurrence. There is currently no reason to make any
distinctions between versions of SPF, because there is currently only
one version. And despite the current RFC, RFC 7208, being an update to
RFC 4408, SPF stayed at version 1 ever since. This commit alters both
the specification text and the message tags.

marc-vanderwal committed:
Zone11: look at name servers, not just their IPs

Amend the message tags and the specification in order to be able to
report the (name, IP) pairs instead of the IP addresses.

This is a breaking change for the following message tags and will
require a database migration script:

 * Z11_DIFFERENT_SPF_POLICIES_FOUND
 * Z11_SPF_MULTIPLE_RECORDS
 * Z11_SPF_SYNTAX_ERROR

marc-vanderwal commented:
I have changed the approach and addressed your review comments. Please re-review.

matsduf commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -67,70 +75,86 @@ same specification.
+    check] for SPF records, then output *[Z11_SPF_SYNTAX_ERROR]* and terminate
+    the test.
+
+11. If the *Child Zone* is the root zone ("."), a [TLD] or a zone within
+    .ARPA, then:
```
```suggestion
11. If the *Child Zone* is the root zone ("."), a [TLD] or a zone under
    .ARPA, then:
```

marc-vanderwal commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
+The root zone ("."), [TLD] zones and zones within .ARPA are treated
+differently because the corresponding domains are not expected to send e-mail.
+Additional discussion on the reason why these zones are unsuitable as e-mail
+domains is provided in the [Zone09 test specification].
```
I’d rather create a seperate PR for clarifying that paragraph in Zone09. Meanwhile, I’ll just rewrite this paragraph here so that the text does not refer to Zone09 anymore.

marc-vanderwal commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
+Z11_INCONSISTENT_SPF_POLICIES    | WARNING |                 | One or more name servers do not publish the same SPF policy as the others.
+Z11_NO_SPF_FOUND                 | NOTICE  | domain          | No SPF policy was found for {domain}.
+Z11_NO_SPF_NON_MAIL_DOMAIN       | INFO    | domain          | No SPF policy was found for {domain}, which is not an e-mail domain.
+Z11_NON_NULL_SPF_NON_MAIL_DOMAIN | NOTICE  | domain          | A non-null SPF policy was found on {domain}, although it is not an e-mail domain.
```
Good suggestion.

marc-vanderwal commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
+Z11_NO_SPF_FOUND                 | NOTICE  | domain          | No SPF policy was found for {domain}.
+Z11_NO_SPF_NON_MAIL_DOMAIN       | INFO    | domain          | No SPF policy was found for {domain}, which is not an e-mail domain.
+Z11_NON_NULL_SPF_NON_MAIL_DOMAIN | NOTICE  | domain          | A non-null SPF policy was found on {domain}, although it is not an e-mail domain.
+Z11_NULL_SPF_NON_MAIL_DOMAIN     | INFO    | domain          | A null SPF policy was found on {domain}, which is not an e-mail domain.
```
Good suggestion, but I’ll slightly reword in order to avoid repeating “which is” twice in the same sentence.

marc-vanderwal commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -67,70 +75,86 @@ same specification.
+   Servers").
 
-4. For each name server in *Name Server IP* do:
+4. For each name server in *Name Servers* do:
```
Okay, I’ll reword as “For each name server IP in *Name Server IP* do:”

matsduf reviewed (dismissed):
Looks fine. Do you plan to create a scenario PR too?

tgreenx reviewed (commented):
LGTM. Just one comment.

marc-vanderwal committed:
Zone11: degrade errors to warnings

Given our definition of what an ERROR is, it is inappropriate to report
Z11_SPF_MULTIPLE_RECORDS, Z11_SYNTAX_ERROR and
Z11_UNABLE_TO_CHECK_FOR_SPF as errors. These tags do not represent
conditions that prevent correct operation for the zone, but at best,
they are conditions that can cause problems in degraded scenarios. We
have to degrade the messages with ERROR level to WARNING level.

See also: zonemaster/zonemaster-engine#1471.

matsduf commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -39,15 +63,18 @@ server.
-Z11_UNABLE_TO_CHECK_FOR_SPF      | ERROR   |                    | None of the zone’s name servers responded with an authoritative response to queries for SPF policies.
+Message Tag                      | Level   | Arguments       | Message ID for message tag
+:--------------------------------|:--------|:----------------|:--------------------------------------------
+Z11_DIFFERENT_SPF_POLICIES_FOUND | NOTICE  | ns_list         | The following name servers returned the same SPF policy, but other name servers returned a different policy. Name servers: {ns_list}.
```
Z11_DIFFERENT_SPF_POLICIES_FOUND is always outputted together with Z11_INCONSISTENT_SPF_POLICIES. The message ID could the be shortened to `The following name servers returned the same SPF policy. Name servers: {ns_list}.`

matsduf commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -1,7 +1,7 @@
 
 ## Test case identifier
-**ZONE11**
+**Zone11**
```
As far as I can see it is always upper case, but we often refer to the test cases with mixed case (except DNSSEC). We should probably keep upper case as @tgreenx suggests.

marc-vanderwal replied:
Okay, I’ll revert that change.

marc-vanderwal committed:
Zone11: editorial updates

matsduf commented on docs/public/specifications/tests/Zone-TP/zone11.md:
```diff
@@ -67,70 +98,91 @@ same specification.
+      .ARPA, then output *[Z11_NO_SPF_NON_MAIL_DOMAIN]* and terminate the
+      test.
+
+   2. Else, output *[Z11_NO_SPF_FOUND]* and terminate the test.
```
It should be stated that the domain is outputted with the tag.

matsduf reviewed (changes requested):
The test case has "ns_list" arguments for some tags, and then there is a chance that the same IP address is used by multiple name server names. It should be stated explicitly that "ns_list" includes all such name/IP pairs.

marc-vanderwal commented:
Step 7 in the test procedure already states:

> For all messages outputted below, if an IP address in *Name Servers* is connected to more than one name server name, then all names should be included with the message tag.

But I found a paragraph in the “Summary” section that was no longer true. I’ve updated it with a similar text. Does this make step 7 in the test procedure redundant?

tgreenx reviewed (approved):
LGTM.

Implementation in https://github.com/zonemaster/zonemaster-engine/pull/1477.

matsduf commented:
@marc-vanderwal, please merge.
